lib/client/api_matrix.es6.js

```diff
@@ -64,49 +91,171 @@ foam.CLASS({
+        if (!browserOptions && !numAvailable) return matrix;
+        for (let webInterface in matrix) {
+          if (!matrix.hasOwnProperty(webInterface)) continue;
+          for (let api in matrix[webInterface]) {
+            if (!Object.prototype.hasOwnProperty
+              .call(matrix[webInterface], api)) {
+                continue;
               }
-              // TODO(markdittmer): ContextualizingDAO should take care of this.
-              return browser.cls_.create(browser, this.__subContext__)
-                  .interfaces.select({
-                    put: (iface) => {
-                      let k0 = iface.interfaceName;
-                      let k1 = iface.apiName;
-                      let k2 = browserId;
-                      // Function and Object's built-in methods can be
-                      // interface name or api name.
-                      // So matrix[k0] is not sufficient to check if
-                      // k0 exits in matrix, use hasOwnProperty instead.
-                      if (!matrix.hasOwnProperty(k0)) {
-                        matrix[k0] = {};
-                      }
-                      // "hasOwnProperty" could also be a apiName,
-                      // thus use Object.prototype.hasOwnProperty
-                      // to ensure we invoke the right function.
-                      if (!Object.prototype.hasOwnProperty
-                        .call(matrix[k0], k1)) {
-                          matrix[k0][k1] = {};
-                        }
-                      matrix[k0][k1][k2] = true;
-                    },
-                    eof: function() {},
-                  });
-                })
-        )).then(() => matrix);
+            let browsers = matrix[webInterface][api];
+            if (this.filterApi_(browsers, numAvailable, browserOptions)) {
+              delete matrix[webInterface][api];
+            }
+          }
+          if (Object.keys(matrix[webInterface]).length === 0) {
+            delete matrix[webInterface];
+          }
+        }
+        return matrix;
+      },
+    },
+    {
+      name: 'filterApi_',
+      documentation: `A helper function to check whether the api should be
+        filtered based on supported browsers, numAvailable and browser
+        options.`,
+      args: [
+        {
+          name: 'supportedBrowsers',
+          documentation: `A list of browsers this api supprts.`,
+        },
+        {
+          name: 'numAvailable',
+          typeName: 'Int/Int Array',
+          documentation: `The same optional argument as
+            options.numAvailable in toMatrix().`,
+        },
+        {
+          name: 'browserOptions',
+          typeName: 'JSON',
+          documentation: `The same optional argument as
+            options.browserOptions in toMatrix().`,
+        },
+      ],
+      returns: {
+        typeName: 'Boolean',
+        documentation: `A boolean value indicates if this API should be
+          filtered.`,
```

Comment:
Indicate clearly what true and false mean, eg. "Returns true if this API matches the filters and should be displayed."